### ISSUE TITLE
Resolving conflicts between secrets in Instance/Binding parameters

### ISSUE THREAD

**Issue by nilebox** [kubernetes-sigs/service-catalog#1090]
This is the extension on top of https://github.com/kubernetes-incubator/service-catalog/pull/1075 to discuss separately.

**Proposal:**

To avoid conflicts for `parametersFrom` (for example, in the case when multiple 
secrets of the same structure need to be passed to the broker), the user can specify 
a `name` field which will produce an extra nesting level.

For example, let's say we have the following structure in a `Secret`:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
data:
  username: YWRtaW4=
  password: MWYyZDFlMmU2N2Rm
```

To inject credentials from two instances of such a `Secret`, we can avoid the 
naming conflict by providing unique names:
```yaml
  ...
  parametersFrom:
    - name: master
      secretRef:
        name: secret1
    - name: slave
      secretRef:
        name: secret2
```
which will result in the following JSON parameters object:
```json
{
  "master": {
    "username": "root",
    "password": "letmein"
  },
  "slave": {
    "username": "foo",
    "password": "bar"
  }
}

```
/cc @duglin

**Comment by nilebox** [comment]
We have discussed this feature with @duglin but didn't come up with a consensus. The options we considered:
- "name" field for nested JSON (as described above)
- "prefix" to add to each top-level parameter name (similar to `envFrom`)
- moving this feature to `parameters` instead of `parametersFrom`

**Comment by duglin** [comment]
I have a feeling this issue might be looking at things from the wrong point of view. This proposal starts with the assumption that we have 2 secrets and we want to send them to a broker and we need to clearly differentiate between the two.  What worries me about solving this problem is that the problem isn't really that we're trying to send 2 secrets. The problem is that the broker wants a master and a slave set of creds and we need to find a way to populate the required data structure - perhaps from 2 secrets.

This may seems like a trivial concern, but I think its important to start from the broker's requirements rather than Kube's since the broker is the component who is asking for the data and will want it in a form natural to its needs.  We may end up with the same solution but I think its an important distinction to make as we design this.

This concern I have actually applies to all of our parameter designs - for simple usecases where there is no nesting/depth then I think we're ok.  But if we do have nesting ( like trying to match the shape of VCAP_SERVICES ) then I wonder how well our design work. And the reason I mentioned VCAP_SERVICES is because I think someone mentioning using the output from a binding as input into a subsequent instance/binding create().  See https://docs.run.pivotal.io/devguide/deploy-apps/environment-variable.html#VCAP-SERVICES  for reference.

So, with that in mind, do we have any real world usecases of service broker parameters that require 2 sets of creds? Or even just creds at all.  I'm most curious to know what the structure looks like and whether the creds will always sit at the top level or if they're nested - because if they're nested more than one level then this solution may not be sufficient.   Of course, these would just be samples and not all possible ways people could expect the data to be sent.

**Comment by duglin** [comment]
I'm not sure I said that clearly, so let me try again from a different perspective ....  can people share some sample broker input/parameter requests? To me this would help define our requirements/usecases since looking at it from the perspective of "how do I send secrets to a broker" isn't really the requirement - its more like "how do I get the secret's data into a form the broker wants" - which means we need to understand what brokers today are expecting.

I'm asking from my Bluemix contacts and have asked some of the Pivotal folks.

**Comment by nilebox** [comment]
> can people share some sample broker input/parameter requests

We don't have any released brokers at the moment, and we can adapt any format which will work best for our use case. I don't expect this feature to be a silver bullet, most probably we'll have to make some preparation of parameters specifically for the broker anyway.

The use case I had in my mind though comes from passing credentials from multiple Bindings as input parameters to an Instance. So each Binding produces a Secret, and if it would be possible to declare Instance parameters by referencing those secrets and without producing a separate Secret, it would be nice.

**Comment by duglin** [comment]
One I got from Matt@pivotal:  https://docs.pivotal.io/spring-cloud-services/1-4/common/config-server/configuring-with-vault.html
Look at the proxy.https.password near the bottom - showing that some brokers do require nesting - even though I do think most probably try to keep the list flat and at the top-level.
